### ISSUE TITLE
E5601: Cannot close window, only floating window would remain

### ISSUE THREAD

**Issue by TornaxO7** [romgrk/nvim-treesitter-context#42]
If I'm inside a function where nvim-treesitter-context gets activated, than I'm getting this error message:
```
E5601: Cannot close window, only floating window would remain
```
So everytime when I want to quit, I need to move out of the nesting in order to close neovim.

**Comment by romgrk** [comment]
Can't reproduce, can you provide the steps you take to get the message?

**Comment by TornaxO7** [comment]
I'm busy currentl I'll try to create a minimal init.vim if I've got the time. But until that I'll close this issue temporary (if it's fine for you).

**Comment by ipod825** [comment]
I can reproduce it. Open any source file, move around until the context floating window shows up. Execute `:close`, the error message shows. Then move around until the context floating window disappears, `:close` works as expected then.

**Comment by romgrk** [comment]
Can't reproduce, with the following steps:

 - ```
    echo 'static
    void my_function(
          int lines,
          char *text
          ) {
    
    }' > test.c
    ```
 - `nvim test.c`
 - `5<c-e>:q<CR>`

**Comment by romgrk** [comment]
Oh ok, I see. The problem is you're using `:close` to quit, but the correct command would be `:quit`. There is nothing we can do here to allow for the behavior you're looking for. I'd suggest using `:quit` (or `ZZ`, or `ZQ`) instead.

**Comment by ipod825** [comment]
`:quit (or ZZ, or ZQ)` behaves just like `close`. Still see the error.

**Comment by romgrk** [comment]
Can't reproduce, please provide an example file with clear steps and a minimal config.

**Comment by TornaxO7** [comment]
Ok, this bug doesn't appear for me anymore. I have absolutely *no* idea, why it doesn't happen... but since @ipod825 has this iccue, I'll reopen this issue.

**Comment by jaekyeom** [comment]
I think I know how to reproduce it.
Try `:quit` when two or more *tabs* are open.

**Comment by TornaxO7** [comment]
https://user-images.githubusercontent.com/50843046/114859828-89d14200-9deb-11eb-8d37-25dab2b8106b.mp4

> I think I know how to reproduce it.
Try :quit when two or more tabs are open.

Yes! It happens for me now!

**Comment by romgrk** [comment]
I can reproduce that. Unfortunately, I don't see a way to fix your use-case, here is the documentation for `:quit`:

```
Quit the current window.  Quit Vim if this is the last window.
```

`:quit` operation order seems to be first close the window, then quit vim. In other words, it's also doing a `:close` and will suffer the same problems as a `:close`. From our end, there is no way to know in advance that the user is going to issue a `:quit` command so we have no way to close the floating window in advance. Mind you, if I have 2 tabpages open and do `:quit`, vim doesn't exit; it only closes one of the tabpages. All that to say, if what you want is to quit vim, the actual command you should be using would be `:qall`. Unless someone as another proposition to fix the issue, I would close this and add a mention on the readme.

**Comment by jaekyeom** [comment]
@romgrk How about providing a wrapper function or command of `:quit`?
I have no knowledge about the internals of nvim-treesitter-context, but at least it should know if the context window is present or not.
If the context window exists and the original window is the last window of the current tab (of course, except for the context window), `winnr('$')` will be `2` and the wrapper can close both the context and original windows. Otherwise, it could just invoke `:quit`.

At the moment, one workaround for users is to use `:tabclose` when it's needed. Caveats I'm aware are 1. it always closes all the windows in the current tab and 2. it fails if the current tab is the only tab.

**Comment by romgrk** [comment]
I'm not opposed, but can you elaborate on why `:qall` doesn't satisfy your use-case?

**Comment by jaekyeom** [comment]
Thanks for considering it.
That's because using `:quit`, I want to close only the current window (which might be the last window in the current tab), not the whole nvim instance.
I thought that a unified wrapper could eliminate the need for manually switching between `:quit` and `:tabclose`, but at the same time, I'm a bit worried that it might complicate such a basic operation (closing a window).
By the way, it looks that there is a related issue for this matter in the neovim project: https://github.com/neovim/neovim/issues/11440

**Comment by romgrk** [comment]
If it's an upstream issue I'd rather fix it upstream, I've opened https://github.com/neovim/neovim/pull/14387.

**Comment by sKyanfIre** [comment]
![Uploading image.png…]()

**Comment by sKyanfIre** [comment]
![image](https://user-images.githubusercontent.com/30263242/119340848-a3f12080-bcc5-11eb-8e16-669474f9601a.png)

**Comment by sKyanfIre** [comment]
The context is saved in Buffer, so `:qa` cannot close it directly

**Comment by luisiacc** [comment]
I did a workaround for this by prepending `G` to the mapping of my quit function: ex:
```
// before
nmap <leader>w :q!

// after
nmap <leader>w G:q!
```

**Comment by TornaxO7** [comment]
I'm also using this, but I'm rather pressing `G` on my own.

**Comment by jaekyeom** [comment]
Fyi, [this comment](https://www.reddit.com/r/neovim/comments/nrz9hp/can_i_close_all_floating_windows_without_closing/h0lg5m1/) suggests a trick for closing all floating windows, which you can use in your mapping.

**Comment by lewis6991** [comment]
This is fixed in Neovim 0.7
